src/runtime/string.cpp

```diff
@@ -672,8 +672,70 @@ static PRIMFN(prim_shell_str) {
   RETURN(String::alloc(runtime.heap, shell_escape(str->c_str())));
 }
 
+static PRIMTYPE(type_substr) {
+  TypeVar result;
+  Data::typeResult.clone(result);
+  result[0].unify(Data::typeString);
+  result[1].unify(Data::typeString);
+  return args.size() == 3 &&
+    args[0]->unify(Data::typeString) &&
+    args[1]->unify(Data::typeInteger) &&
+    args[2]->unify(Data::typeInteger) &&
+    out->unify(result);
+}
+
+// Create a 'Result String String' value on the heap.
+Value *allocate_result(Heap &h, bool is_pass, const std::string &msg) {
+    h.reserve(reserve_result() + String::reserve(msg.size()));
+    return claim_result(h, is_pass, String::claim(h, msg));
+}
+
+// This operates on utf8 runes rather than bytes, to match prim_explode.
+static PRIMFN(prim_substr) {
+  EXPECT(3);
+  STRING(input_str, 0);
+  INTEGER_MPZ(pos_arg, 1);
+  INTEGER_MPZ(len_arg, 2);
+
+  int start_pos = mpz_get_si(pos_arg);
+  int length = mpz_get_si(len_arg);
+  if (start_pos < 0 || length < 1) {
+    RETURN(allocate_result(runtime.heap, false, "substr: invalid position or length argument"));
+  }
+
+  std::string new_substr;
+  new_substr.reserve(length);
+
+  // number of variable width runes seen
+  int rune_count = 0;
+  // width of the rune
+  int rune_num_bytes = 0;
+
+  for (const char *ptr = input_str->c_str(); *ptr; ptr += rune_num_bytes) {
+    uint32_t rune;
+    rune_num_bytes = pop_utf8(&rune, ptr);
+    if (rune_num_bytes < 1) {
+      rune_num_bytes = 1;
```

Comment:
You probably don't want to push_utf8 here, as that would change the encoding.
Probably, you should just copy the byte over.

Reply:
In this scenario, an actual `Fail` might be appropriate. I'm not keen on making substr return `Result`, though, as you often want to use it in a larger expression, and needing a `require` adds a lot of friction, there.

Reply:
I made the function have well defined behaviour in this case, by carefully defining a codepoint.